From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Luke=20Sern=C3=A9?= <luke.serne@solconmail.nl>
Date: Fri, 10 Jul 2026 19:10:00 +0200
Subject: [PATCH] 9363: Decompiler: Increase maximum type recovery steps

Decompiler: Increase maximum type recovery steps This allows
 properly finding the right structure field in deeply nested structures.
 Previously, the cutoff was 7, and now it is 16.
---
 Ghidra/Features/Decompiler/src/decompile/cpp/coreaction.cc | 6 ++++--
 1 file changed, 4 insertions(+), 2 deletions(-)

diff --git a/Ghidra/Features/Decompiler/src/decompile/cpp/coreaction.cc b/Ghidra/Features/Decompiler/src/decompile/cpp/coreaction.cc
index 20411eddc5..3b122d414f 100644
--- a/Ghidra/Features/Decompiler/src/decompile/cpp/coreaction.cc
+++ b/Ghidra/Features/Decompiler/src/decompile/cpp/coreaction.cc
@@ -5857,6 +5857,8 @@ void ActionInferTypes::propagateAcrossReturns(Funcdata &data)
   }
 }
 
+#define MAX_ITERS_TYPE_RECOVERY 16       // This constant arrived at empirically
+
 int4 ActionInferTypes::apply(Funcdata &data)
 
 {
@@ -5873,8 +5875,8 @@ int4 ActionInferTypes::apply(Funcdata &data)
     data.getArch()->printDebug(s.str());
   }
 #endif
-  if (localcount >= 7) {       // This constant arrived at empirically
-    if (localcount == 7) {
+  if (localcount >= MAX_ITERS_TYPE_RECOVERY) {
+    if (localcount == MAX_ITERS_TYPE_RECOVERY) {
       data.warningHeader("Type propagation algorithm not settling");
       data.setTypeRecoveryExceeded();
       localcount += 1;
-- 
2.45.1

